## Bài 1: Tạo Tensor cơ bản

In [6]:
import torch

a = torch.tensor(
[[1, 2, 3],
[4, 5, 6]],
dtype=torch.int32
)
print(a)

tensor([[1, 2, 3],
        [4, 5, 6]], dtype=torch.int32)


#### Giải thích 
- Kích thước của Tensor là 2x3=6.
- Kiểu dữ liệu của Tensor a là int

## Bài 2: Tạo các Tensor thường dùng

In [3]:
import torch
a = torch.rand(3, 4)
b = torch.randn(3, 4)
c = torch.zeros(2, 3)
d = torch.ones(2, 3)
e = torch.eye(4)
f = torch.full((5, 5), 10)
print(a)
print(b)
print(c)
print(d)
print(e)
print(f)

tensor([[0.2716, 0.0341, 0.9214, 0.5536],
        [0.1682, 0.2372, 0.0565, 0.5955],
        [0.6678, 0.0059, 0.7232, 0.7444]])
tensor([[ 0.6421, -1.3363,  0.1764,  0.4771],
        [-0.8373, -0.0700,  1.8225, -0.5663],
        [ 0.1519, -1.0734, -1.2492,  0.4401]])
tensor([[0., 0., 0.],
        [0., 0., 0.]])
tensor([[1., 1., 1.],
        [1., 1., 1.]])
tensor([[1., 0., 0., 0.],
        [0., 1., 0., 0.],
        [0., 0., 1., 0.],
        [0., 0., 0., 1.]])
tensor([[10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10]])


#### Nhận xét

- `torch.rand()` tạo các giá trị ngẫu nhiên theo phân phối đều trong khoảng `[0, 1)`.
- `torch.randn()` tạo các giá trị ngẫu nhiên theo phân phối chuẩn với trung bình gần 0 và độ lệch chuẩn gần 1.
- Vì vậy, `torch.rand()` chỉ cho giá trị không âm trong `[0, 1)`, còn `torch.randn()` có thể cho cả giá trị âm và dương.
- Mục đích thêm `f = torch.full((5,5),10)` để làm đầy đủ cho tensor.

## Bài 3: Kiểm tra thông tin Tensor

In [4]:
import torch
a = torch.zeros(2, 3, 4)

print("Tensor:")
print(a)
print("Shape:", a.shape)
print("Size:", a.size())
print("Number of dimensions:", a.ndim)
print("Data type:", a.dtype)

Tensor:
tensor([[[0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.]],

        [[0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.]]])
Shape: torch.Size([2, 3, 4])
Size: torch.Size([2, 3, 4])
Number of dimensions: 3
Data type: torch.float32


#### Giải thích ý nghĩa của torch.Size([2,3,4])
- Chiều 1: có 2 phần tử
- Chiều 2: có 3 phần tử
- Chiều 3: có 4 phần tử
- `Vậy a có dạng 2x3x4=24` phần tử, tất cả phần tử đều bằng 0. 



## Bài 4. Slicing và Indexing

In [ ]:
import torch
a = torch.randn(3, 4, 5)
print("Original tensor:")
print(a)
print("\nTensor a[1]:")
print(a[1])
print("\nTensor a[1:, 2:4]:")
print(a[1:, 2:4])

Original tensor:
tensor([[[ 0.1157, -1.0295, -0.7091, -1.4609,  0.7907],
         [-0.4948,  0.7378,  0.9093, -0.3913,  2.2236],
         [ 1.8157, -1.0778,  0.3947,  0.3892,  0.4362],
         [-1.1805, -0.6514, -0.0294, -0.8011,  1.1038]],

        [[ 1.4245, -0.6620,  1.6494, -1.1918,  0.2962],
         [ 0.5678,  0.0035,  0.3379,  2.0635,  1.3949],
         [ 0.3614, -0.2736, -0.6554, -1.7323, -0.2380],
         [ 1.0640,  0.0375,  0.9831,  0.7486,  0.7150]],

        [[-0.5060, -0.9003, -0.8447, -0.5782, -0.5743],
         [-2.3309, -0.5844,  0.2787,  1.0397, -1.4116],
         [-0.5445,  1.9222,  1.0075,  0.8291, -0.3827],
         [-1.2950, -2.1506, -0.7087, -2.0582,  1.9775]]])

Tensor a[1]:
tensor([[ 1.4245, -0.6620,  1.6494, -1.1918,  0.2962],
        [ 0.5678,  0.0035,  0.3379,  2.0635,  1.3949],
        [ 0.3614, -0.2736, -0.6554, -1.7323, -0.2380],
        [ 1.0640,  0.0375,  0.9831,  0.7486,  0.7150]])
Shape of a[1]: torch.Size([4, 5])

Tensor a[1:, 2:4]:
tensor([[[ 0.361

#### Yêu cầu:

- Shape của `a[1]` là: torch.Size([4,5]).
- Shape của `a[1:, 2:4]` là: torch.Size([2, 2, 5]).
- Shape of a[:, :, -2:]: torch.Size([3, 4, 2]). Hai phần tử cuối cùng trên chiều thứ ba: `[-2.0582, 1.9775]`

## Bài 5. Reshape và Flatten

In [8]:
import torch
a = torch.randn(3, 4)
print("Original:")
print(a)
print("\nFlatten:")
print(a.ravel())
print("\nReshape to 3 x 2 x 2:")
print(a.reshape(3, 2, 2))

Original:
tensor([[ 0.1021,  1.1481,  0.1752, -0.9851],
        [-0.6169,  0.3260,  0.5171, -0.6098],
        [-2.2642, -0.0488,  0.5929,  0.3377]])

Flatten:
tensor([ 0.1021,  1.1481,  0.1752, -0.9851, -0.6169,  0.3260,  0.5171, -0.6098,
        -2.2642, -0.0488,  0.5929,  0.3377])

Reshape to 3 x 2 x 2:
tensor([[[ 0.1021,  1.1481],
         [ 0.1752, -0.9851]],

        [[-0.6169,  0.3260],
         [ 0.5171, -0.6098]],

        [[-2.2642, -0.0488],
         [ 0.5929,  0.3377]]])


In [9]:
print("\nReshape to 2 x 6:")
print(a.reshape(2, 6))


Reshape to 2 x 6:
tensor([[ 0.1021,  1.1481,  0.1752, -0.9851, -0.6169,  0.3260],
        [ 0.5171, -0.6098, -2.2642, -0.0488,  0.5929,  0.3377]])


#### Yêu cầu:
- Số phần tử trước và sau khi reshape không thay đổi, chỉ thay đổi từ 3 hàng 4 cột -> 6 hàng 2 cột.
- Đã `a.reshape(2,6) cho Tensor.

## Bài 6. Phép toán và nhân ma trận

In [12]:
import torch
a = torch.randn(2, 3)
b = torch.randn(2, 3)
print("a:")
print(a)
print("b:")
print(b)
print("\na+b:")
print(a + b)
print("\na / b:")
print(a / b)
print("\na^2:")
print(a ** 2)
print("\nMatrix multiplication:")
print(a @ b.T)

a:
tensor([[ 0.9694, -0.8778, -0.8472],
        [-0.2307, -0.7082, -0.8635]])
b:
tensor([[ 1.0785, -0.1407, -0.1577],
        [ 0.8167, -0.4614,  0.2760]])

a+b:
tensor([[ 2.0479, -1.0185, -1.0049],
        [ 0.5860, -1.1696, -0.5875]])

a / b:
tensor([[ 0.8989,  6.2380,  5.3707],
        [-0.2825,  1.5350, -3.1288]])

a^2:
tensor([[0.9398, 0.7705, 0.7177],
        [0.0532, 0.5016, 0.7456]])

Matrix multiplication:
tensor([[ 1.3026,  0.9629],
        [-0.0130, -0.1000]])


In [ ]:
## them a*b để so sánh sự khác nhau
print("\na*b:")
print(a * b)
## in ra shape của a@b.T
print("\nShape:", (a@b.T).shape)


a*b:
tensor([[ 1.0455,  0.1235,  0.1336],
        [-0.1884,  0.3268, -0.2383]])

Shape: torch.Size([2, 2])


#### Yêu cầu:
- Sau khi cho randn 2  tensor a và b và thực hiện phép toán, phân biệt với a và b có shape (2, 3):
a * b: nhân từng phần tử tương ứng, kết quả có shape (2, 3).
a @ b.T: nhân ma trận. Vì b.T có shape (3, 2), nên: (2, 3) @ (3, 2) = (2, 2).
- Shape của kết quả a @ b.T: torch.Size([2,2]).

## Bài 7: Các hàm thống kê trên Tensor

In [16]:
import torch
a = torch.randn(3, 4)
print(a)
print("\nMean:")
print(torch.mean(a, dim=0))
print("\nStandard deviation:")
print(torch.std(a, dim=0))
print("\nCumulative sum:")
print(torch.cumsum(a, dim=0))

tensor([[ 0.4466, -0.2788,  1.0306, -0.0017],
        [ 1.3445,  0.5067,  1.0112,  0.1261],
        [ 0.2874,  0.2882,  0.7385, -1.1817]])

Mean:
tensor([ 0.6928,  0.1721,  0.9268, -0.3524])

Standard deviation:
tensor([0.5700, 0.4054, 0.1634, 0.7210])

Cumulative sum:
tensor([[ 4.4655e-01, -2.7879e-01,  1.0306e+00, -1.6734e-03],
        [ 1.7911e+00,  2.2795e-01,  2.0418e+00,  1.2447e-01],
        [ 2.0785e+00,  5.1619e-01,  2.7803e+00, -1.0572e+00]])


In [18]:
print("\nMean:")
print(torch.mean(a, dim=1))
print("\nStandard deviation:")
print(torch.std(a, dim=1))
print("\nCumulative sum:")
print(torch.cumsum(a, dim=1))


Mean:
tensor([0.2992, 0.7472, 0.0331])

Standard deviation:
tensor([0.5719, 0.5385, 0.8372])

Cumulative sum:
tensor([[0.4466, 0.1678, 1.1984, 1.1967],
        [1.3445, 1.8513, 2.8625, 2.9886],
        [0.2874, 0.5756, 1.3141, 0.1324]])


#### Yêu cầu:
- `dim=0`: thao tác theo chiều thứ nhất, tức là tính toán dọc theo các hàng. Với `a` có shape `(3, 4)`, kết quả của `mean` và `std` có 4 giá trị, mỗi giá trị tương ứng với một cột. `cumsum` cũng có shape `(3, 4)` và cộng dồn từ hàng trên xuống hàng dưới.
- `dim=1`: thao tác theo chiều thứ hai, tức là tính toán ngang theo từng hàng. Kết quả của `mean` và `std` có 3 giá trị, mỗi giá trị tương ứng với một hàng. `cumsum` cũng có shape `(3, 4)` và cộng dồn từ cột trái sang cột phải.
- Vì vậy, thay `dim=0` bằng `dim=1` làm thay đổi hướng tính toán và kết quả của `mean`, `std`, nhưng không làm thay đổi shape của `cumsum`.

## Bài 8. Autograd cơ bản

In [ ]:
import torch

x = torch.tensor(3.6, requires_grad=True)
y = x * x
y.backward()

print("Ham 1: y = x^2")
print("x =", x)
print("y =", y)
print("Dao ham thu cong dy/dx = 2x =", 2 * x.item())
print("x.grad =", x.grad)

## doi ham y=x^2 thanh y = 3*x^2 + 2*x + 1
x_new = torch.tensor(3.6, requires_grad=True)
y_new = 3 * x_new**2 + 2 * x_new + 1
y_new.backward()

print("\nHam 2: y = 3*x^2 + 2*x + 1")
print("x =", x_new)
print("y =", y_new)
print("Dao ham thu cong dy/dx = 6x + 2 =", 6 * x_new.item() + 2)
print("x_new.grad =", x_new.grad)

Ham 1: y = x^2
x = tensor(3.6000, requires_grad=True)
y = tensor(12.9600, grad_fn=<MulBackward0>)
Dao ham thu cong dy/dx = 2x = 7.199999809265137
x.grad = tensor(7.2000)

Ham 2: y = 3*x^2 + 2*x + 1
x = tensor(3.6000, requires_grad=True)
y = tensor(47.0800, grad_fn=<AddBackward0>)
Dao ham thu cong dy/dx = 6x + 2 = 23.59999942779541
x_new.grad = tensor(23.6000)


### Giải thích

- `requires_grad = True` yêu cầu PyTorch theo dõi các phép tính liên quan đến x để tính đạo hàm.
- `y = x * x` tạo hàm `y = x^2`.
- `y.backward()` áp dụng quy tắc lan truyền ngược để tính `dy/dx`.
- `x.grad` lưu gradient của `y` theo `x`.

Với `y = x^2`:

- Đạo hàm: `dy/dx =2*x`.
- Tại `x = 3.6`: `dy/dx = 2 * 3.6 = 7.2`.
- Kết quả này giống `x.grad = 7.2`.

Với `y = 3*x^2 + 2*x + 1`:

- Đạo hàm: `dy/dx = 6*x + 2`.
- Tại `x = 3.6`: `dy/dx = 6 * 3.6 + 2 = 23.6`.
- Kết quả này trùng với `x_new.grad = 23.6`.

## Bài 9. Gradient Descent tìm hệ số đa thức

In [32]:
import numpy as np
import torch

polynomial = np.poly1d([1, 2, 3])
N = 20
X = np.random.randn(N, 1) * 5
Y = polynomial(X)
XX = np.hstack([
    X * X,
    X,
    np.ones_like(X)
])

w = torch.randn(3, 1, requires_grad=True)
x = torch.tensor(XX, dtype=torch.float32)
y = torch.tensor(Y, dtype=torch.float32)
optimizer = torch.optim.NAdam([w], lr=0.01)

target_coefficients = torch.tensor([[1.0], [2.0], [3.0]])
print("Initial coefficients:")
print(w.detach().flatten())

for epoch in range(5000):
    optimizer.zero_grad()
    y_pred = x @ w
    mse = torch.mean(torch.square(y - y_pred))
    mse.backward()
    optimizer.step()

print("Learned coefficients [x^2, x, 1]:")
print(w.detach().flatten())
print("Target coefficients:")
print(target_coefficients.flatten())
print("Coefficient errors:")
print((w.detach() - target_coefficients).flatten())
print("Final MSE:", mse.item())

Initial coefficients:
tensor([-0.3934,  0.7321, -0.2731])
Learned coefficients [x^2, x, 1]:
tensor([1.0000, 2.0000, 2.9999])
Target coefficients:
tensor([1., 2., 3.])
Coefficient errors:
tensor([ 1.4305e-06,  4.7684e-07, -8.8930e-05])
Final MSE: 4.582710655398614e-09


### So sánh và giải thích

- `w` gồm ba hệ số theo thứ tự `[x^2, x, 1]`. Vì vậy, nghiệm đúng cần tìm là `[1, 2, 3]`.
- Sau khi huấn luyện, các hệ số học được xấp xỉ `[1.0000, 2.0000, 3.0001]`, rất gần với `[1, 2, 3]`. Sai số nhỏ là do quá trình tối ưu số và số vòng lặp hữu hạn.
- `optimizer.zero_grad()` xóa gradient của vòng lặp trước. Nếu không xóa, gradient mới sẽ bị cộng dồn vào gradient cũ.
- `mse.backward()` dùng Autograd để tính gradient của hàm mất mát MSE theo `w`.
- `optimizer.step()` dùng các gradient vừa tính để cập nhật `w` theo thuật toán NAdam, giúp MSE giảm dần.
- `y_pred = x @ w` dự đoán giá trị của đa thức; `mse` đo độ khác nhau giữa dự đoán và dữ liệu thật.

## Bài 10. Giải hệ phương trình bằng Autograd

In [35]:
import random
import torch

A = torch.tensor(random.random(), requires_grad=True)
B = torch.tensor(random.random(), requires_grad=True)
C = torch.tensor(random.random(), requires_grad=True)
D = torch.tensor(random.random(), requires_grad=True)
EPOCHS = 2000
optimizer = torch.optim.NAdam([A, B, C, D], lr=0.01)

for _ in range(EPOCHS):
    y1 = A + B - 9
    y2 = C - D - 1
    y3 = A + C - 8
    y4 = B - D - 2
    sqerr = y1**2 + y2**2 + y3**2 + y4**2

    optimizer.zero_grad()
    sqerr.backward()
    optimizer.step()

with torch.no_grad():
    residuals = torch.tensor([
        A + B - 9,
        C - D - 1,
        A + C - 8,
        B - D - 2
    ])
    final_sqerr = torch.sum(residuals**2)

print("A =", A.item())
print("B =", B.item())
print("C =", C.item())
print("D =", D.item())
print("\nKiem tra cac phuong trinh:")
print("A + B =", (A + B).item(), "(muc tieu: 9)")
print("C - D =", (C - D).item(), "(muc tieu: 1)")
print("A + C =", (A + C).item(), "(muc tieu: 8)")
print("B - D =", (B - D).item(), "(muc tieu: 2)")
print("Sai so tung phuong trinh:", residuals)
print("Tong sai so binh phuong:", final_sqerr.item())

A = 4.212620258331299
B = 4.787327766418457
C = 3.7873613834381104
D = 2.7873423099517822

Kiem tra cac phuong trinh:
A + B = 8.999948501586914 (muc tieu: 9)
C - D = 1.0000190734863281 (muc tieu: 1)
A + C = 7.999981880187988 (muc tieu: 8)
B - D = 1.9999854564666748 (muc tieu: 2)
Sai so tung phuong trinh: tensor([-5.1498e-05,  1.9073e-05, -1.8120e-05, -1.4544e-05])
Tong sai so binh phuong: 3.5557263800001238e-09


### Nhận xét

- Sau khi tối ưu, bốn phương trình đều được thỏa mãn gần đúng; tổng sai số bình phương rất nhỏ nên nghiệm số là phù hợp.
- Hệ phương trình có nhiều nghiệm vì phương trình thứ tư phụ thuộc vào ba phương trình đầu:
  `B - D = (9 - A) - (7 - A) = 2`.
- Đặt `A = t` tùy ý, ta có họ nghiệm:
  `A = t`, `B = 9 - t`, `C = 8 - t`, `D = 7 - t`.
- Do đó, bài toán có một tham số tự do và không có một nghiệm duy nhất. NAdam chỉ tìm một nghiệm cụ thể tùy thuộc vào giá trị khởi tạo và quá trình tối ưu.

## Bài 11. Tổng hợp Tensor, Autograd và tối ưu đa thức

In [36]:
import torch

# Tao Tensor X co 24 phan tu va reshape theo ba cach
X = torch.arange(24)
X_4x6 = X.reshape(4, 6)
X_2x12 = X.reshape(2, 12)
X_2x3x4 = X.reshape(2, 3, 4)
print("X reshape 4 x 6:")
print(X_4x6)
print("\nX reshape 2 x 12:")
print(X_2x12)
print("\nX reshape 2 x 3 x 4:")
print(X_2x3x4)

# Phep toan va thong ke theo cot
A = torch.randn(3, 4)
B = torch.randn(3, 4)
print("\nA + B:")
print(A + B)
print("\nA * B:")
print(A * B)
print("\nA / B:")
print(A / B)
print("\nTrung binh theo cot:")
print(torch.mean(A, dim=0))
print("\nDo lech chuan theo cot:")
print(torch.std(A, dim=0))

# Autograd cho y = 2*x^2 + 5*x + 3 tai x = 2
x_value = torch.tensor(2.0, requires_grad=True)
y_value = 2 * x_value**2 + 5 * x_value + 3
y_value.backward()
manual_gradient = 4 * 2 + 5
print("\nAutograd:")
print("y =", y_value.item())
print("dy/dx bang Autograd =", x_value.grad.item())
print("dy/dx thu cong =", manual_gradient)

# Hoc lai he so cua y = 2*x^2 - 3*x + 5 bang NAdam
polynomial = torch.tensor([2.0, -3.0, 5.0])
torch.manual_seed(0)
x_train = torch.linspace(-5, 5, 40).reshape(-1, 1)
y_train = polynomial[0] * x_train**2 + polynomial[1] * x_train + polynomial[2]
features = torch.cat([x_train**2, x_train, torch.ones_like(x_train)], dim=1)
coefficients = torch.randn(3, 1, requires_grad=True)
optimizer = torch.optim.NAdam([coefficients], lr=0.01)

for _ in range(5000):
    optimizer.zero_grad()
    prediction = features @ coefficients
    loss = torch.mean((y_train - prediction)**2)
    loss.backward()
    optimizer.step()

print("\nHe so hoc duoc [x^2, x, 1]:")
print(coefficients.detach().flatten())
print("He so muc tieu:")
print(polynomial)
print("Sai so he so:")
print((coefficients.detach().flatten() - polynomial))
print("MSE cuoi:", loss.item())

X reshape 4 x 6:
tensor([[ 0,  1,  2,  3,  4,  5],
        [ 6,  7,  8,  9, 10, 11],
        [12, 13, 14, 15, 16, 17],
        [18, 19, 20, 21, 22, 23]])

X reshape 2 x 12:
tensor([[ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11],
        [12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23]])

X reshape 2 x 3 x 4:
tensor([[[ 0,  1,  2,  3],
         [ 4,  5,  6,  7],
         [ 8,  9, 10, 11]],

        [[12, 13, 14, 15],
         [16, 17, 18, 19],
         [20, 21, 22, 23]]])

A + B:
tensor([[ 1.6394, -1.6653,  0.4018,  1.5245],
        [-0.2033,  1.1638,  2.0070,  1.0222],
        [ 2.5475, -0.0165,  0.9871, -1.5609]])

A * B:
tensor([[ 0.2477, -1.0717, -0.1314,  0.0960],
        [-0.0926,  0.3322,  0.6632,  0.2593],
        [ 1.4585, -0.7769, -0.4262,  0.4736]])

A / B:
tensor([[ 8.7346, -0.2294, -2.8811, 22.1531],
        [-0.5188,  0.7579,  3.8108,  0.8411],
        [ 0.5176, -1.0189, -4.0388,  0.3590]])

Trung binh theo cot:
tensor([0.8530, 0.0360, 1.1724, 0.5044])

Do lech chuan the

### Kết quả

- `X` có 24 phần tử nên có thể reshape thành `4 x 6`, `2 x 12` hoặc `2 x 3 x 4`; tổng số phần tử vẫn là 24.
- Với `A` và `B` có shape `(3, 4)`, các phép `A + B`, `A * B` và `A / B` đều thực hiện theo từng phần tử và cho shape `(3, 4)`.
- `torch.mean(A, dim=0)` và `torch.std(A, dim=0)` tính theo cột, nên mỗi kết quả có 4 giá trị.
- Với `y = 2*x^2 + 5*x + 3`, đạo hàm là `dy/dx = 4x + 5`. Tại `x = 2`, kết quả thủ công và Autograd đều bằng 13.
- Với đa thức `y = 2*x^2 - 3*x + 5`, các hệ số học được theo thứ tự `[x^2, x, 1]` xấp xỉ `[2, -3, 5]`.